pkg/epp/flowcontrol/controller/controller.go

```diff
@@ -0,0 +1,478 @@
+	defer ticker.Stop()
+
+	fc.reconcileProcessors() // Initial reconciliation
+
+	for {
+		select {
+		case <-ctx.Done():
+			fc.shutdown()
+			return
+		case <-ticker.C():
+			fc.reconcileProcessors()
+		}
+	}
+}
+
+// EnqueueAndWait is the primary, synchronous entry point to the Flow Control system. It submits a request and blocks
+// until the request reaches a terminal outcome (dispatched, rejected, or evicted).
+//
+// # Design Rationale: The Synchronous Model
+//
+// This blocking model is deliberately chosen for its simplicity and robustness, especially in the context of Envoy
+// External Processing (`ext_proc`), which operates on a stream-based protocol.
+//
+//   - `ext_proc` Alignment: A single goroutine typically manages the stream for a given HTTP request.
+//     `EnqueueAndWait` fits this perfectly: the request-handling goroutine calls it, blocks, and upon return, has a
+//     definitive outcome to act upon.
+//   - Simplified State Management: The state of a "waiting" request is implicitly managed by the blocked goroutine's
+//     stack and its `context.Context`. The system only needs to signal this specific goroutine to unblock it.
+//   - Direct Backpressure: If queues are full, `EnqueueAndWait` returns an error immediately, providing direct
+//     backpressure to the caller.
+func (fc *FlowController) EnqueueAndWait(req types.FlowControlRequest) (types.QueueOutcome, error) {
+	if req == nil {
+		return types.QueueOutcomeRejectedOther, fmt.Errorf("%w: %w", types.ErrRejected, types.ErrNilRequest)
+	}
+	effectiveTTL := req.InitialEffectiveTTL()
+	if effectiveTTL <= 0 {
+		effectiveTTL = fc.config.DefaultRequestTTL
+	}
+	enqueueTime := fc.clock.Now()
+
+	for {
+		if !fc.isRunning.Load() {
+			return types.QueueOutcomeRejectedOther, fmt.Errorf("%w: %w", types.ErrRejected, types.ErrFlowControllerNotRunning)
+		}
+
+		// We must create a fresh `FlowItem` on each attempt since finalization is idempotent.
+		// However, it we use the original, preserved `enqueueTime`.
+		item := internal.NewItem(req, effectiveTTL, enqueueTime)
+		if outcome, err := fc.distributeRequest(item); err != nil {
+			return outcome, fmt.Errorf("%w: %w", types.ErrRejected, err)
+		}
+
+		finalState := <-item.Done() // finalization handles monitoring request context cancellation and TTL expiry
+		if errors.Is(finalState.Err, contracts.ErrShardDraining) {
+			fc.logger.V(logutil.DEBUG).Info("Shard is draining, retrying request", "requestID", req.ID())
+			// Benign race with the chosen `contracts.RegistryShard` becoming Draining post selection but before the item was
+			// enqueued into its respective `contracts.ManagedQueue`. Simply try again.
+			continue
+		}
+
+		return finalState.Outcome, finalState.Err
+	}
+}
+
+// distributeRequest selects the optimal shard processor for a given item and attempts to submit it.
+//
+// # Architectural Deep Dive: Achieving Emergent Fairness with Data Parallelism
```

Comment:
This design justification should be documented somewhere for future contributors and maintenance. I scoped it to the relevant function that implements the algorithm, but I am not convinced this should be documented here as it is quite verbose.